src/Microsoft.AspNetCore.Mvc.Razor.Precompilation.Tools/Internal/PrecompileDispatchCommand.cs

```diff
@@ -0,0 +1,153 @@
+using System.Linq;
+using System.Reflection;
+using Microsoft.DotNet.Cli.Utils;
+using Microsoft.DotNet.InternalAbstractions;
+using Microsoft.DotNet.ProjectModel;
+using Microsoft.Extensions.CommandLineUtils;
+using Microsoft.Extensions.Internal;
+using NuGet.Frameworks;
+
+namespace Microsoft.AspNetCore.Mvc.Razor.Precompilation.Internal
+{
+    public class PrecompileDispatchCommand : PrecompileCommandBase
+    {
+        private CommandOption NoBuildOption { get; set; }
+
+        protected override void Configure(CommandLineApplication app)
+        {
+            base.Configure(app);
+            NoBuildOption = app.Option(
+                "--no-build",
+                "Do not build project before compiling views.",
+                CommandOptionType.NoValue);
+        }
+
+        protected override int ExecuteCore()
+        {
+            if (!NoBuildOption.HasValue())
+            {
+                var exitCode = BuildProject();
+                if (exitCode != 0)
+                {
+                    return exitCode;
+                }
+            }
+
+            var dispatchArgs = new List<string>
+            {
+                ProjectPath,
+                "--framework",
+                TargetFramework.ToString(),
+                "--configuration",
+                Configuration,
+            };
+
+            string outputPath = null;
+            if (OutputPathOption.HasValue())
+            {
+                outputPath = OutputPathOption.Value();
+
+                dispatchArgs.Add("--output");
+                dispatchArgs.Add(outputPath);
+            }
+
+            if (ConfigureCompilationType.HasValue())
+            {
+                dispatchArgs.Add("--configure-compilation-type");
+                dispatchArgs.Add(ConfigureCompilationType.Value());
+            }
+
+            if (ContentRootOption.HasValue())
+            {
+                dispatchArgs.Add("--content-root");
+                dispatchArgs.Add(ContentRootOption.Value());
+            }
+
+            if (GeneratePdbOption.HasValue())
+            {
+                dispatchArgs.Add("--generate-pdbs");
+            }
+
+            var toolName = typeof(Precompilation.Program).GetTypeInfo().Assembly.GetName().Name;
+            var dispatchCommand = DotnetToolDispatcher.CreateDispatchCommand(
+                dispatchArgs,
+                TargetFramework,
+                Configuration,
+                outputPath: outputPath,
+                buildBasePath: null,
+                projectDirectory: ProjectPath,
+                toolName: toolName);
+
+            var commandExitCode = dispatchCommand
+                .ForwardStdErr(Console.Error)
+                .ForwardStdOut(Console.Out)
+                .Execute()
+                .ExitCode;
+
+            var buildOutputPath = GetBuildOutputPath();
+            var updatedBinary = Directory.EnumerateFiles(buildOutputPath, $"*{ModifiedAssemblyExtension}").FirstOrDefault();
```

Comment:
In theory there's only one we generate. Probably needs some to be hardened against stale binaries.